Expression, Intrinsic operators

Hywan opened the pull request:
Address #72.
Address https://github.com/tagua-vm/parser/issues/14.

#### Specification

https://github.com/php/php-langspec/blob/master/spec/10-expressions.md#intrinsics

#### Progression

 * [x] `array-intrinsic`,
 * [x] `empty-intrinsic`,
 * [x] `eval-intrinsic`,
 * [x] `exit-intrinsic`,
 * [x] `isset-intrinsic`,
 * [x] `print-intrinsic`.

Hywan committed:
Expression: Add the `intrinsic_eval` rule.

Hywan committed:
Test: Add an invalid case for `instrinsic_eval`.

Hywan commented:
@nikic `exit(2)` is valid regarding the spec and Zend Engine implementation. `exit 2` is valid regarding the spec but not Zend Engine implementation. Is an issue from the spec?

For the record, the extract from the current spec:

```
  exit-intrinsic:
    exit  expressionopt
    exit  (  expressionopt  )
    die   expressionopt
    die   (   expressionopt )
```

nikic commented:
@Hywan Yeah, spec bug. Should be

```
  exit-intrinsic:
    exit
    exit  (  expressionopt  )
    die
    die   (   expressionopt )
```

Hywan commented:
@nikic Would you like me to open a PR?

nikic commented:
@Hywan Yeah, would be great!

Hywan committed:
Expression: Add the `intrinsic_exit` rule.

nikic commented on source/rules/expressions/primaries.rs:
```diff
@@ -184,6 +200,71 @@ fn empty_mapper<'a>(expression: Expression<'a>) -> StdResult<Expression<'a>, ()>
+                    return Err(Error::Code(ErrorKind::Custom(IntrinsicError::ReservedExitCode as u32)));
+                } else if code > 255 {
+                    return Err(Error::Code(ErrorKind::Custom(IntrinsicError::OutOfRangeExitCode as u32)));
+                }
```
I don't think it makes sense to do these checks at this point -- after all the argument can also be an expression, so this is something the runtime needs to deal with. (Though I don't think PHP does anything special here at all.)

Hywan replied:
Yup I know. I would like to make a double checks. We can check this constraint at parse-time if this is an integer (reals are considered as strings in Zend Engine, strange, I would expect to be casted into integer… anyway), and another check at runtime.

Most of the time, I feel like integers are used directly as constants. So this is a first security.

Hywan replied:
I feel like if this can be checked at parse-time easily, we should do it to avoid any “hidden error” (at least a test case should cover this instruction at runtime). Of course, `$a = 12; exit($a);` could be checked at runtime (or with a static or concolic approach for instance), but if we can avoid a malformed instruction easily, I feel we should do it. Thoughts?

Hywan commented:
@nikic Thank you for following the project and having a close look at what we are doing :-]. Very much appreciated and precious!

Hywan committed:
Expression: Add the `intrinsic_isset` rule.

Hywan committed:
Expression: `Unset` is a vector of expressions.

`Expression::Unset(Vec<Variable<'a>>)` becomes
`Expression::Unset(Vec<Expression<'a>>)`.

Hywan committed:
Documentation: Complete `Expression::Isset`.

Hywan committed:
Expression: `intrinsic_isset` is an operator.

`intrinsic_isset` was declared as a `intrinsic_construct` instead of an
`intrinsic_operator`.

Hywan committed:
Expression: Add the `intrinsic_print` rule.

Hywan committed:
Expression: Add the `intrisic_array` rule.

Hywan commented:
Related PR on the spec: https://github.com/php/php-langspec/pull/176, https://github.com/php/php-langspec/pull/175 and https://github.com/php/php-langspec/pull/174.

Hywan commented:
Need to support the long `array(…)` form, and by reference.

Hywan committed:
Test: Add a recursive case for `intrisic_array`.

Hywan committed:
Expression: More skip tokens in `intrisic_array`.

Hywan committed:
Expression: Factorise `[` in `intrinsic_array`.

The `[` token is parsed twice in case of an non-empty array. This
situation can be avoided by factorising the matching of `[` with a
parent `preceded!` macro.

Hywan committed:
Expression: Support long array syntax.

`array(…)` is strictly equivalent to `[…]`. The former is the long
syntax, the latter is the short syntax.

Hywan committed:
Test: Add cases for trailing comma in `…ic_array`.

Hywan committed:
Expression: Support value by reference in an array.
